core_backend/app/llm_call/parse_input.py

```diff
@@ -2,75 +2,196 @@
 These are functions that can be used to parse the input questions.
 """
 
+from functools import wraps
+from typing import Any, Callable, Tuple
+
 from ..configs.llm_prompts import (
     PARAPHRASE_INPUT,
     TRANSLATE_INPUT,
     IdentifiedLanguage,
     SafetyClassification,
 )
-from ..schemas import UserQueryRefined
+from ..schemas import ResultState, UserQueryRefined, UserQueryResponse
 from .utils import _ask_llm
 
 
-def classify_safety(query_text: str) -> SafetyClassification:
+def classify_safety(func: Callable) -> Callable:
+    """
+    Decorator to classify the safety of the question.
+    """
+
+    @wraps(func)
+    async def wrapper(
+        question: UserQueryRefined,
+        response: UserQueryResponse,
+        *args: Any,
+        **kwargs: Any,
+    ) -> UserQueryResponse:
+        """
+        Wrapper function to classify the safety of the question.
+        """
+        question, response = _classify_safety(question, response)
+        response = await func(question, response, *args, **kwargs)
+        return response
+
+    return wrapper
+
+
+def _classify_safety(
+    question: UserQueryRefined, response: UserQueryResponse
+) -> Tuple[UserQueryRefined, UserQueryResponse]:
     """
     Classifies the safety of the question.
     """
-    return getattr(
-        SafetyClassification,
-        _ask_llm(query_text, SafetyClassification.get_prompt()),
-    )
+    if response.state != ResultState.ERROR:
```

Comment:
In all the rails, if there is an error already in the response, from maybe a previous rail, just return the response and don't do any other action.

Reply:
P.S: Monads are a pretty cool way to do this but alas no simple way to do that in Python. Why functional programming is so awesome.